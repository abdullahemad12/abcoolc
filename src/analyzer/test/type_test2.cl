class Main
{
	main() : Int { if 1 then 1 else 2 fi};
	meth() : Int { if 1 = 2 then 1 else 4 fi };
	meth1() : B { if 1 < 2 then new C else new D fi};
	meth2() : A { if 1 < 2 then new C else new D fi };

	meth3() : String 
	{
		{
			1;
			new Main;
			new A; 
			new B; 
			"hello";
		}
	};
	meth4() : String 
	{
		{
			1;
			new Main;
			new A; 
			new B; 
			1;
		}
	};
		
	x : String <- "hello";
	let_meth() : A
	{
		{
			let x : Bool <- true in x;
			x <- "hello world";
			let x : Int <- 2, b : B <- new B, d : D <- new D in if x = 2 then b else d fi;
		}
	};
	let_meth1() : Int
	{
		{
			let x : Bool <- true in x;
			x <- "hello world";
			let x : Int <- 2, b : B <- new B, d : D <- new D in if x = 2 then x <- 2 else x <- 3 fi;
		}
	};
	let_meth3(x : Bool) : A
	{
		{
			x = true;
			let x : Bool <- true in x;
			x <- "hello world";
			let x : Int in x + 3;
			let x : Int <- 2, b : B <- new B, d : D <- new D in if x = 2 then new Main else d fi;
		}
	};				
};


class A {
	get_int() : Int { 1 };
	case_meth(x : A) : B 
	{
		case x.get_int() of 
			x : Int => x + 3;
			x : B => 4;
			y : C => y.case_meth(y);
		esac
	};
	
	case_meth1(x : A) : A 
	{
		case x.get_int() of 
			x : Int => {x <- x + 1; new B;};
			x : B => x;
			y : C => {y.case_meth(y); y;};
			y : D => y;
		esac
	};
};

class B inherits A
{
	loop_meth(x : Int, n : Int) : Int
	{
		let i : Int <- 0 in while(i <= n) loop x <- x + 1 pool
	};

	loop_meth1(x : Int, n : Int) : Object
	{
		let i : Int <- 0 in while(i <= n) loop x <- x + 1 pool
	};

	is_void_meth() : Bool
	{
		isvoid new C.loop_meth1(1, 2)
	};
	

	is_void_meth1() : Int
	{
		isvoid new C.loop_meth1(1, 2)
	};
	
	not_void() : Int
	{
		Not 3
	};

	not_meth(x : Bool) : Bool 
	{
		not x
	};

	compare_meth(x : Int, y : Int) : Int
	{
		if x < y then if x <= y then x else y fi else y fi
	};

	
	compare_meth(x : Int, y : Int) : Int
	{
		if x < y then if x <= y then (x <= y) + 1 else y fi else (y < x) + 2 fi
	};

	neg_meth(x : Int) : Int
	{
		~x
	};

	neg_meth2(x : Bool) : Bool
	{
		~x
	};

	arith(x : Int, y : Int, z : Int, w : String) : Int 
	{
		{
			x + w; 
			x + y; 
			x / w;
			w * y;
			z * x;
			y - x;
		}
	};

	zz : String; 
	ww : Bool; 
	equal(x : Int, y : Int, z : String, w : Bool) : Bool
	{
		{
			x = y;
			x = z;
			x = w;
			z = zz;
			z = w;
			w = ww;
		}
	};
};

class C inherits B
{
	x : B <- self;
	y : Int <- y;
};

class D inherits A 
{

};
