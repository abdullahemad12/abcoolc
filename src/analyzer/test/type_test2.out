#1
_program
  #1
  _class
    Main
    Object
    "test/type_test2.cl"
    (
    #3
    _method
      main
      Int
      #3
      _cond
        #3
        _int
          1
        : _no_type
        #3
        _int
          1
        : _no_type
        #3
        _int
          2
        : _no_type
      : _no_type
    #4
    _method
      meth
      Int
      #4
      _cond
        #4
        _eq
          #4
          _int
            1
          : _no_type
          #4
          _int
            2
          : _no_type
        : _no_type
        #4
        _int
          1
        : _no_type
        #4
        _int
          4
        : _no_type
      : _no_type
    #5
    _method
      meth1
      B
      #5
      _cond
        #5
        _lt
          #5
          _int
            1
          : _no_type
          #5
          _int
            2
          : _no_type
        : _no_type
        #5
        _new
          C
        : _no_type
        #5
        _new
          D
        : _no_type
      : _no_type
    #6
    _method
      meth2
      A
      #6
      _cond
        #6
        _lt
          #6
          _int
            1
          : _no_type
          #6
          _int
            2
          : _no_type
        : _no_type
        #6
        _new
          C
        : _no_type
        #6
        _new
          D
        : _no_type
      : _no_type
    #8
    _method
      meth3
      String
      #10
      _block
        #11
        _int
          1
        : _no_type
        #12
        _new
          Main
        : _no_type
        #13
        _new
          A
        : _no_type
        #14
        _new
          B
        : _no_type
        #15
        _string
          "hello"
        : _no_type
      : _no_type
    #18
    _method
      meth4
      String
      #20
      _block
        #21
        _int
          1
        : _no_type
        #22
        _new
          Main
        : _no_type
        #23
        _new
          A
        : _no_type
        #24
        _new
          B
        : _no_type
        #25
        _int
          1
        : _no_type
      : _no_type
    #29
    _attr
      x
      String
      #29
      _string
        "hello"
      : _no_type
    #30
    _method
      let_meth
      A
      #32
      _block
        #33
        _let
          x
          Bool
          #33
          _bool
            1
          : _no_type
          #33
          _object
            x
          : _no_type
        : _no_type
        #34
        _assign
          x
          #34
          _string
            "hello world"
          : _no_type
        : _no_type
        #35
        _let
          x
          Int
          #35
          _int
            2
          : _no_type
          #35
          _let
            b
            B
            #35
            _new
              B
            : _no_type
            #35
            _let
              d
              D
              #35
              _new
                D
              : _no_type
              #35
              _cond
                #35
                _eq
                  #35
                  _object
                    x
                  : _no_type
                  #35
                  _int
                    2
                  : _no_type
                : _no_type
                #35
                _object
                  b
                : _no_type
                #35
                _object
                  d
                : _no_type
              : _no_type
            : _no_type
          : _no_type
        : _no_type
      : _no_type
    #38
    _method
      let_meth1
      Int
      #40
      _block
        #41
        _let
          x
          Bool
          #41
          _bool
            1
          : _no_type
          #41
          _object
            x
          : _no_type
        : _no_type
        #42
        _assign
          x
          #42
          _string
            "hello world"
          : _no_type
        : _no_type
        #43
        _let
          x
          Int
          #43
          _int
            2
          : _no_type
          #43
          _let
            b
            B
            #43
            _new
              B
            : _no_type
            #43
            _let
              d
              D
              #43
              _new
                D
              : _no_type
              #43
              _cond
                #43
                _eq
                  #43
                  _object
                    x
                  : _no_type
                  #43
                  _int
                    2
                  : _no_type
                : _no_type
                #43
                _assign
                  x
                  #43
                  _int
                    2
                  : _no_type
                : _no_type
                #43
                _assign
                  x
                  #43
                  _int
                    3
                  : _no_type
                : _no_type
              : _no_type
            : _no_type
          : _no_type
        : _no_type
      : _no_type
    #46
    _method
      let_meth3
      #46
      _formal
        x
        Bool
      A
      #48
      _block
        #49
        _eq
          #49
          _object
            x
          : _no_type
          #49
          _bool
            1
          : _no_type
        : _no_type
        #50
        _let
          x
          Bool
          #50
          _bool
            1
          : _no_type
          #50
          _object
            x
          : _no_type
        : _no_type
        #51
        _assign
          x
          #51
          _string
            "hello world"
          : _no_type
        : _no_type
        #52
        _let
          x
          Int
          #52
          _no_expr
          : _no_type
          #52
          _plus
            #52
            _object
              x
            : _no_type
            #52
            _int
              3
            : _no_type
          : _no_type
        : _no_type
        #53
        _let
          x
          Int
          #53
          _int
            2
          : _no_type
          #53
          _let
            b
            B
            #53
            _new
              B
            : _no_type
            #53
            _let
              d
              D
              #53
              _new
                D
              : _no_type
              #53
              _cond
                #53
                _eq
                  #53
                  _object
                    x
                  : _no_type
                  #53
                  _int
                    2
                  : _no_type
                : _no_type
                #53
                _new
                  Main
                : _no_type
                #53
                _object
                  d
                : _no_type
              : _no_type
            : _no_type
          : _no_type
        : _no_type
      : _no_type
    )
  #59
  _class
    A
    Object
    "test/type_test2.cl"
    (
    #60
    _method
      get_int
      Int
      #60
      _int
        1
      : _no_type
    #61
    _method
      case_meth
      #61
      _formal
        x
        A
      B
      #63
      _typcase
        #63
        _dispatch
          #63
          _object
            x
          : _no_type
          get_int
          (
          )
        : _no_type
        #64
        _branch
          x
          Int
          #64
          _plus
            #64
            _object
              x
            : _no_type
            #64
            _int
              3
            : _no_type
          : _no_type
        #65
        _branch
          x
          B
          #65
          _int
            4
          : _no_type
        #66
        _branch
          y
          C
          #66
          _dispatch
            #66
            _object
              y
            : _no_type
            case_meth
            (
            #66
            _object
              y
            : _no_type
            )
          : _no_type
      : _no_type
    #70
    _method
      case_meth1
      #70
      _formal
        x
        A
      A
      #72
      _typcase
        #72
        _dispatch
          #72
          _object
            x
          : _no_type
          get_int
          (
          )
        : _no_type
        #73
        _branch
          x
          Int
          #73
          _block
            #73
            _assign
              x
              #73
              _plus
                #73
                _object
                  x
                : _no_type
                #73
                _int
                  1
                : _no_type
              : _no_type
            : _no_type
            #73
            _new
              B
            : _no_type
          : _no_type
        #74
        _branch
          x
          B
          #74
          _object
            x
          : _no_type
        #75
        _branch
          y
          C
          #75
          _block
            #75
            _dispatch
              #75
              _object
                y
              : _no_type
              case_meth
              (
              #75
              _object
                y
              : _no_type
              )
            : _no_type
            #75
            _object
              y
            : _no_type
          : _no_type
        #76
        _branch
          y
          D
          #76
          _object
            y
          : _no_type
      : _no_type
    )
  #81
  _class
    B
    A
    "test/type_test2.cl"
    (
    #83
    _method
      loop_meth
      #83
      _formal
        x
        Int
      #83
      _formal
        n
        Int
      Int
      #85
      _let
        i
        Int
        #85
        _int
          0
        : _no_type
        #85
        _loop
          #85
          _leq
            #85
            _object
              i
            : _no_type
            #85
            _object
              n
            : _no_type
          : _no_type
          #85
          _assign
            x
            #85
            _plus
              #85
              _object
                x
              : _no_type
              #85
              _int
                1
              : _no_type
            : _no_type
          : _no_type
        : _no_type
      : _no_type
    #88
    _method
      loop_meth1
      #88
      _formal
        x
        Int
      #88
      _formal
        n
        Int
      Object
      #90
      _let
        i
        Int
        #90
        _int
          0
        : _no_type
        #90
        _loop
          #90
          _leq
            #90
            _object
              i
            : _no_type
            #90
            _object
              n
            : _no_type
          : _no_type
          #90
          _assign
            x
            #90
            _plus
              #90
              _object
                x
              : _no_type
              #90
              _int
                1
              : _no_type
            : _no_type
          : _no_type
        : _no_type
      : _no_type
    #93
    _method
      is_void_meth
      Bool
      #95
      _isvoid
        #95
        _dispatch
          #95
          _new
            C
          : _no_type
          loop_meth1
          (
          #95
          _int
            1
          : _no_type
          #95
          _int
            2
          : _no_type
          )
        : _no_type
      : _no_type
    #99
    _method
      is_void_meth1
      Int
      #101
      _isvoid
        #101
        _dispatch
          #101
          _new
            C
          : _no_type
          loop_meth1
          (
          #101
          _int
            1
          : _no_type
          #101
          _int
            2
          : _no_type
          )
        : _no_type
      : _no_type
    #104
    _method
      not_void
      Int
      #106
      _comp
        #106
        _int
          3
        : _no_type
      : _no_type
    )
  #110
  _class
    C
    B
    "test/type_test2.cl"
    (
    )
  #115
  _class
    D
    A
    "test/type_test2.cl"
    (
    )
